### PR TITLE
Functional Test improvements for Snupkg and File Not Found scenarios

### LINKED ISSUE #8837: Add Missing PushCommand Functional Tests for snupkg and FileNotFound scenarios
## Details about Problem
PushCommandTests currently miss testing several features of Push involving snupkgs and FileNotFound errors. 

These tests are needed so that there is existing test code that can be modified in the soon to be resolved bug fix: https://github.com/NuGet/Home/issues/8148

### PR DESCRIPTION
**Pull request by donnie-msft**
## Bug
Missing test coverage relevant to upcoming work on https://github.com/NuGet/Home/issues/8148.

Fixes: https://github.com/NuGet/Home/issues/8837 
Regression: No  
Details: Adding missing test coverage and did some refactoring of Test utilities.

### CODE REVIEW

**Comment by donnie-msft** [comment]
Bad commit merge - will re-request reviews once fixed

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalFolderUtilityTests.cs [hunk 1]
```diff
@@ -1030,6 +1032,65 @@ public void LocalFolderUtility_GetPackagesByIdV3NotFoundMissingDir()
...
+        {
+            //Arrange
+            List<string> testList = new List<string>();
+            string packagePath = string.Empty;
+
+            //Act
+            try
+            {
+                LocalFolderUtility.EnsurePackageFileExists(packagePath, testList);
+                Assert.True(false, "Did not throw an exception as expected.");
```
Comment (kartheekp-ms) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalFolderUtilityTests.cs]: Isn't this line of code is unreachable if expected behavior is to throw exception ?. May be use `Assert.Throws` directly instead of assert in the exception handler.

Comment (donnie-msft) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalFolderUtilityTests.cs]: Correct. The line is designed to fail if the code doesn't throw an exception. I didn't know about Assert.Throws and can take a look at that since it sounds cleaner.

Comment (donnie-msft) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalFolderUtilityTests.cs]: This look good, now ?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs [hunk 2]
```diff
@@ -17,6 +17,7 @@
 using NuGet.Versioning;
```
Comment (dominoFire) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs:20]: Copyright at the beginning of the the file

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs [hunk 3]
```diff
@@ -557,6 +570,18 @@ public static void AddPublishResource(JObject index, MockServer publishServer)
             array.Add(resource);
         }
 
+        public static void AddPublishSymbolsResource(JObject index, MockServer publishServer)
```
Comment (dominoFire) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs:576]: Where do you use this method?

Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs:576]: Nevermind. Ignore this comment

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MockServer.cs [hunk 4]
```diff
@@ -188,6 +188,11 @@ public static void SavePushedPackage(HttpListenerRequest r, string outputFileNam
             }
         }
 
+        public static bool PushedPackageExists(string outputFileName)
```
Comment (dominoFire) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MockServer.cs]: Where is this method used? I didn't find any references to it by using Find All References in VS.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs [hunk 5]
```diff
@@ -245,6 +256,512 @@ public void PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds()
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// When pushing *.Snupkg, (no skip duplicate) a 409 Conflict is returned and halts the remaining symbols push.
+        /// TODO: bug fixes will come from https://github.com/NuGet/Home/issues/8148
+        /// </summary>
+        [Fact]
+        public void PushCommand_Server_Snupkg_WildcardFindsMatchingSnupkgs_Conflict()
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Isn't this an opposite scenario of what NuGet.org does?

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs [hunk 6]
```diff
@@ -245,6 +256,512 @@ public void PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds()
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// When pushing *.Snupkg with SkipDuplicate, a 409 Conflict is ignored and the remaining symbols push proceeds.
+        /// TODO: bug fixes will come from https://github.com/NuGet/Home/issues/8148
+        /// </summary>
+        [Fact]
+        public void PushCommand_Server_Snupkg_WildcardFindsMatchingSnupkgs_SkipDuplicate()
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Same here,

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs [hunk 7]
```diff
@@ -245,6 +256,512 @@ public void PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds()
...
+            }
+        }
+
+
+        /// <summary>
+        /// When pushing snupkg with SkipDuplicate, a 409 Conflict is ignored and any message from the server is shown appropriately.
+        /// TODO: bug fixes will come from https://github.com/NuGet/Home/issues/8148
+        /// </summary>
+        [Fact]
+        public void PushCommand_Server_Snupkg_FilenameSnupkgExists_SkipDuplicate_ServerMessage()
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: It's not clear to me from the speclet (it'd be nice to put the spec in the designs repo for future tracking) whether the nuget.org behavior is changing, or all of it is client side.

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs [hunk 8]
```diff
@@ -70,10 +71,18 @@ public SimpleTestPackageContext()
         public string RuntimeJson { get; set; }
 
         public bool IsSymbolPackage { get; set; }
+        public bool IsSnupkgPackage { get; set; } = false;
```
Comment (nkolev92) [inline test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs]: You don't need to explicitly set a default.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs [hunk 9]
```diff
@@ -245,6 +256,512 @@ public void PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds()
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// When pushing *.nupkg to a symbol server, but no snupkgs are selected with that wildcard, there is (TODO: not) a FileNotFound error about snupkgs.
+        /// TODO: bug fixes will come from https://github.com/NuGet/Home/issues/8148
+        /// </summary>
+        [Fact]
+        public void PushCommand_Server_Nupkg_WildcardSnupkgDoesNotExist_FileNotFoundError()
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: How do you explicitly push to a symbol server? 

Isn't this the same behavior for any globbing pattern?

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs [hunk 10]
```diff
@@ -245,6 +256,512 @@ public void PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds()
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// When pushing a nupkg by filename to a Symbol Server with a matching snupkg, a 409 Conflict halts the push.
+        /// TODO: bug fixes will come from https://github.com/NuGet/Home/issues/8148
+        /// </summary>
+        [Fact]
+        public void PushCommand_Server_Nupkg_FilenameSnupkgExists_Conflict()
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: The name doesn't really communicate the intent here clearly. 

Is the snupkg push failing because the equivalent nupkg push did?

Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: How is this different from the below test?

**Review by nkolev92 — COMMENTED** [review]
I like that you are covering so many scenarios. 

But it's a little bit hard to follow what the current behavior is. 

The TODOs will change are confusing. These tests should validate the current behavior. The notion of a future PR is irrelevant for the reader.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 5 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Nope. NuGet.org will return conflicts when a snupkg is being validated.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 6 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: I don't understand what you're asking?

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 7 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: This test has nothing to do with NuGet.org. These are NuGet Client Functional Tests, which are testing how the Client handles file not found, and pushing snupkgs with corresponding nupkgs.

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs (hunk 8 above)
Comment (donnie-msft) [reply test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs]: This is a new property, and none of the existing code knows about its existence. Rather than modify all of those places, I prefer simply defaulting it.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs [hunk 11]
```diff
@@ -245,6 +256,512 @@ public void PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds()
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// When pushing *.Nupkg with SkipDuplicate, a 409 Conflict is ignored and the secondary symbols push proceeds.
+        /// TODO: bug fixes will come from https://github.com/NuGet/Home/issues/8148
+        /// </summary>
+        [Fact]
+        public void PushCommand_Server_Nupkg_WildcardFindsMatchingSnupkgs_SkipDuplicate()
```
Comment (dominoFire) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: The following tests looks similar:

- `PushCommand_Server_Nupkg_WildcardFindsMatchingSnupkgs_SkipDuplicate`
- `PushCommand_Server_Snupkg_WildcardFindsMatchingSnupkgs_SkipDuplicate`

Can those tests generalized using a `[Theory]`? Does the behavior will be different once NuGet/HOme#8148 is solved?

The same comment also applies for all the tests.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 9 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: The server needs to have [SymbolPackagePublish](https://github.com/NuGet/NuGet.Client/blob/f256c56155a5b32d54ef3d8cf2bc28d197862df4/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs#L578) as I setup the Mock server to do.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 10 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: The pattern is `PushCommand_Server_<what is being pushed>_<filename/wildcard><does the file exist?>_<Server Response>` and I'm open to suggested changes to the pattern, which follows similarly to the other Functional Tests.

Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: The difference is in bold: PushCommand_Server_Nupkg_**WildcardFindsMatching**Snupkgs_Conflict

Meaning that this test is using a Wildcard rather than a Filename, and the wildcard finds matching files, but the server returns a Conflict.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 11 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: They are similar, but very different code paths occur when pushing Nupkgs vs. Snupkgs, directly.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs [hunk 12]
```diff
@@ -245,6 +256,512 @@ public void PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds()
...
+                    // Act
+                    var result = CommandRunner.Run(
+                        nuget,
+                        packageDirectory,
+                        $"push {snupkgToPush} -Source {sourceName} -Timeout 110",
+                        waitForExit: true,
+                        timeOutInMilliseconds: 120000); // 120 seconds
+
+                    // Assert
+                    server.Stop();
```
Comment (dominoFire) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Why do you need to stop the server here? Can you rely on the `using` language construct to stop the server?

**Review by dominoFire — APPROVED** [review]
Looks good. Let's keep focus on the solution for https://github.com/NuGet/Home/issues/8148

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs (hunk 8 above)
Comment (rrelyea) [reply test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs]: bool properties, by default, will default to false. that may be @nkolev92's point?

Comment (donnie-msft) [reply test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs]: Sure, but since it's important to the behavior, it should be there. Also, if that's the case then this comment should be a "nit:"

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 12 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Was following the pattern from original tests. Actually, there's other tests where I could stop it sooner, maybe even dispose it sooner, so thanks for pointing that out.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 9 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: And no, the behavior is different for globbing snupkgs vs globbing nupkgs. Which is the point of many of these new tests.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 5 above)
Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Ok, makes sense. So NuGet.org might return either 409 or 201 depending on how much it's been since the last push.

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs (hunk 8 above)
Comment (nkolev92) [reply test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs]: Discussed offline. 
No need for default values.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 9 above)
Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Is *.nupkg supposed to be selecting snupkgs? 

If there's only snupkgs, and we do *.nupkg, would we push the snupkgs?

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 10 above)
Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: The fact that nuget.org might 409/201 even after a push wasn't clear. 
Given that I can see why you want the 2 tests.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 6 above)
Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: I couldn't follow what the behavior should be based on the spec. 
You cleared it up somewhere in the other comments.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 7 above)
Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: I couldn't follow what the behavior should be based on the spec.
You cleared it up somewhere in the other comments.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 9 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Client behavior is to try to push the equivalent snupkg of whatever nupkg was pushed. 
*.nupkg -> *.snupkg
filename.nupkg -> filename.snupkg

No, if there's only snupkgs, then *.nupkg is an Error, and `PushCommand_Server_Nupkg_WildcardSnupkgDoesNotExist_FileNotFoundError`. was intended to test that. When I do my next PR, I'll clear this up and split it into 2 tests by creating one just for *.nupkg with no files found. Thanks for pointing it out.

Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: So this is testing the buggy behavior right now?

Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: These tests are testing existing behavior. We have determined that it's incorrect behavior, so I will be fixing it in another Issue.

Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Regarding the *.nupkg not found, I added 2 additional tests for that.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 12 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: Refactored all tests to dispose the MockServer sooner (before Assertions)

**Review by nkolev92 — APPROVED** [review]
